### ISSUE TITLE
Problemas with Entity Attachment To Email

### ISSUE THREAD

**Issue by romeusantiago** [demianrasko/Dynamics-365-Workflow-Tools#196]
I'm trying to create an e-mail and when I select the "Retrieve ActivityMimeAttachment" or the "Select Most Recent Distinct Files" as "False" I'm getting an error.

Message error:

"Plugin Trace:
[msdyncrmWorkflowTools: msdyncrmWorkflowTools.Class.EntityAttachmentToEmail]
[msdyncrmWorkflowTools (1.0.58.0): Entity Attachment To Email]

Load CRM Service from context --- OK
ParentObjectTypeCode=10098--ParentId=8e8defea-1275-ea11-a811-000d3ac17d61
FetchXML: <fetch version='1.0' output-format='xml-platform' mapping='logical' distinct='false'
                                 >
                      <entity name='annotation'>
                        <attribute name='filename' />
                        <attribute name='annotationid' />
                        <attribute name='subject' />
                        <attribute name='documentbody' />
                        <attribute name='mimetype' />
                        <order attribute='createdon' descending='true' />
                        <filter type='and'>
                          <condition attribute='filename' operator='like' value='%.%' />
                          <condition attribute='isdocument' operator='eq' value='1' />
                          <condition attribute='objectid' operator='eq' value='8e8defea-1275-ea11-a811-000d3ac17d61 ' />
                        </filter>
                      </entity>
                    </fetch> 
Entities Count: 1 

Error Message:

Unhandled exception: 
Exception type: Microsoft.Crm.CrmException
Message: An unexpected error occurred from ISV code. (ErrorType = ClientError) Unexpected exception from plug-in (Execute):  msdyncrmWorkflowTools.Class.EntityAttachmentToEmail: System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.Crm.Sandbox.SandboxCodeUnit.ProcessException(Exception originalException, IExecutionContext context, SandboxClient client, SandboxCallTracker callTracker, Boolean isSafeToRetry, DateTime performanceExecutionStartTime, SandboxTracker tracker, Guid parentExecutionId, CrmException& crmException, String& assemblyContents)
   at Microsoft.Crm.Sandbox.SandboxCodeUnit.<>c__DisplayClass26_0.<Execute>b__0()
   at Microsoft.PowerApps.CoreFramework.ActivityLoggerExtensions.Execute(ILogger logger, EventId eventId, ActivityType activityType, Action action, IEnumerable`1 additionalCustomProperties)
   at Microsoft.Xrm.Telemetry.XrmTelemetryExtensions.Execute(ILogger logger, XrmTelemetryActivityType activityType, Action action)
   at Microsoft.Xrm.RemotePlugin.CrmProvider.RemotePlugin.CDSExecute(IServiceProvider serviceProvider)
   at Microsoft.Crm.Workflow.Services.ProxyCustomActivity.Execute(CodeActivityContext executionContext)
-- End stack trace --
"

When the options are selected as "True" the workflow runs but do not recover the attachment from entity.

Can you check it and help me? I might be messing with something but I'd like you to take a look, please.

My environment is a Dynamics 365 (Online) - Version 1710 (9.1.0.1984) with the version 1.0.58.0.

![Entity Attachment Error](https://user-images.githubusercontent.com/15317054/86840670-64075780-c079-11ea-8145-b5912e7ae0d8.jpg)

![Entity Attachment Error 2](https://user-images.githubusercontent.com/15317054/86840673-649fee00-c079-11ea-8a34-236a6a8ce56a.jpg)

![Entity Attachment Error 3](https://user-images.githubusercontent.com/15317054/86840674-65388480-c079-11ea-8f98-d9050cc64a72.jpg)

Thanks in advance and best regards.  ;)
Romeu Santiago.

**Comment by demianrasko** [comment]
Hi @romeusantiago You shoul use the "Retrieve ActivityMimeAttachment=true" only if the "Main Record URL" type is email or appointment.
Please try on this, and you tell me.

thanks!

**Comment by romeusantiago** [comment]
> Hi @romeusantiago You shoul use the "Retrieve ActivityMimeAttachment=true" only if the "Main Record URL" type is email or appointment.
> Please try on this, and you tell me.
> 
> thanks!

Tks for the reply, @demianrasko! ;) Should I update the install to the version 1.0.59.0 or it gonna with the 1.0.58.0?
Best regards and thanks in advance.

**Comment by demianrasko** [comment]
Hi @romeusantiago it should work in both versions. No changes in the new version on this point.

**Comment by romeusantiago** [comment]
Hi @demianrasko, I've tried again today and still have problems.

The step "msdyncrmWorkflowTools (1.0.58.0):Entity Attachment To Email" is giving error and I've fixed and did what you instruct me to do.

But I had the same error and message. Am I missing something?

Just to explain, I'm trying to get the Attachment related to a custom entity and add it on an e-mail.

Just that!

Gonna attach  the error screen.
![Error 1 - 01-07-2020](https://user-images.githubusercontent.com/15317054/87492329-b4a22600-c620-11ea-99da-063924ccd028.jpg)


"Plugin Trace:

[msdyncrmWorkflowTools: msdyncrmWorkflowTools.Class.EntityAttachmentToEmail]
[msdyncrmWorkflowTools (1.0.58.0): Entity Attachment To Email]

Load CRM Service from context --- OK
ParentObjectTypeCode=10098--ParentId=8e8defea-1275-ea11-a811-000d3ac17d61
FetchXML: <fetch version='1.0' output-format='xml-platform' mapping='logical' distinct='false'
                                 >
                      <entity name='annotation'>
                        <attribute name='filename' />
                        <attribute name='annotationid' />
                        <attribute name='subject' />
                        <attribute name='documentbody' />
                        <attribute name='mimetype' />
                        <order attribute='createdon' descending='true' />
                        <filter type='and'>
                          <condition attribute='filename' operator='like' value='%.%' />
                          <condition attribute='isdocument' operator='eq' value='1' />
                          <condition attribute='objectid' operator='eq' value='8e8defea-1275-ea11-a811-000d3ac17d61 ' />
                        </filter>
                      </entity>
                    </fetch> 
Entities Count: 1 



Error Message:

Unhandled exception: 
Exception type: Microsoft.Crm.CrmException
Message: An unexpected error occurred from ISV code. (ErrorType = ClientError) Unexpected exception from plug-in (Execute):  msdyncrmWorkflowTools.Class.EntityAttachmentToEmail: System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.Crm.Sandbox.SandboxCodeUnit.ProcessException(Exception originalException, IExecutionContext context, SandboxClient client, SandboxCallTracker callTracker, Boolean isSafeToRetry, DateTime performanceExecutionStartTime, SandboxTracker tracker, Guid parentExecutionId, CrmException& crmException, String& assemblyContents)
   at Microsoft.Crm.Sandbox.SandboxCodeUnit.<>c__DisplayClass26_0.<Execute>b__0()
   at Microsoft.PowerApps.CoreFramework.ActivityLoggerExtensions.Execute(ILogger logger, EventId eventId, ActivityType activityType, Action action, IEnumerable`1 additionalCustomProperties)
   at Microsoft.Xrm.Telemetry.XrmTelemetryExtensions.Execute(ILogger logger, XrmTelemetryActivityType activityType, Action action)
   at Microsoft.Xrm.RemotePlugin.CrmProvider.RemotePlugin.CDSExecute(IServiceProvider serviceProvider)
   at Microsoft.Crm.Workflow.Services.ProxyCustomActivity.Execute(CodeActivityContext executionContext)
-- End stack trace -- "


Again I thank you for the help. ;)
